### PR TITLE
test upi generation

### PR DESCRIPTION
**Pull request by jyotigautam**
Why?
[test] To test the UPI generating function for products  
[bug] Fix redirection on variant deletion

This change addresses the need by:
1. Mocking UPI generating function i.e, gen_nano_id and make it return repeated nano ids to test the UPI duplication scenario. 
2. Redirect to parent product's edit page on variant deletion.
3. Handle nil case while fetching taxon using get_taxon function in taxon controller.

[delivers #163288827](https://www.pivotaltracker.com/story/show/163288827)

### CODE REVIEW

File: apps/snitch_core/test/data/model/product_test.exs [hunk 1]
```diff
@@ -152,21 +155,15 @@ defmodule Snitch.Data.Model.ProductTest do
...
-      end
+  test "upi generation for a product", %{valid_params: vp} do
+    with_mock GenNanoid, gen_nano_id: fn -> NanoidMock.gen_nano_id() end do
+      NanoidMock.start_link(0)
+      {:ok, product1} = Product.create(vp)
+      vp = %{vp | name: "latest test product"}
+      {:ok, product2} = Product.create(vp)
+      assert product1.upi == "A0C"
+      refute product1.upi == product2.upi
+      NanoidMock.stop()
```
Comment (ashish173) [inline apps/snitch_core/test/data/model/product_test.exs:169]: @jyotigautam Please group the instructions and leave spaces in between.

File: mix.lock [hunk 2]
```diff
@@ -79,11 +79,13 @@
   "makeup_elixir": {:hex, :makeup_elixir, "0.10.0", "0f09c2ddf352887a956d84f8f7e702111122ca32fbbc84c2f0569b8b65cbf7fa", [:mix], [{:makeup, "~> 0.5.5", [hex: :makeup, repo: "hexpm", optional: false]}], "hexpm"},
   "mariaex": {:hex, :mariaex, "0.8.4", "5dd42a600c3949ec020cfac162a815115c9e9e406abffc1b14ffdc611d6f84bc", [:mix], [{:db_connection, "~> 1.1", [hex: :db_connection, repo: "hexpm", optional: false]}, {:decimal, "~> 1.0", [hex: :decimal, repo: "hexpm", optional: false]}, {:poison, ">= 0.0.0", [hex: :poison, repo: "hexpm", optional: true]}], "hexpm"},
   "maybe": {:hex, :maybe, "1.0.0", "65311dd7e16659579116666b268d03d7e1d1b3da8776c81a6b199de7177b43d6", [:mix], [], "hexpm"},
+  "meck": {:hex, :meck, "0.8.13", "ffedb39f99b0b99703b8601c6f17c7f76313ee12de6b646e671e3188401f7866", [:rebar3], [], "hexpm"},
```
Comment (ashish173) [inline mix.lock:82]: @jyotigautam is this intended?

File: apps/snitch_core/test/data/model/product_test.exs (hunk 1 above)
Comment (jyotigautam) [reply apps/snitch_core/test/data/model/product_test.exs:169]: Done the formatting.

File: mix.lock (hunk 2 above)
Comment (jyotigautam) [reply mix.lock:82]: No, this was not done intentionally. Meck is a dependency of mock.

Comment (ashish173) [reply mix.lock:82]: okay cool! my bad

**Review by ashish173 — APPROVED** [review]

**Pull request merged**